### ISSUE TITLE
Unrolled lists pop twice after push

### ISSUE THREAD

**Issue by mrjbq7** [factor/factor#2729]
```factor
IN: scratchpad USE: unrolled-lists

IN: scratchpad <unrolled-list> 1 over push-back
               dup pop-back . dup pop-back .
1
0
```

Should be the same as a pop with no push:

```factor
IN: scratchpad <unrolled-list> pop-back
empty-deque


Type :help for debugging help.
```

**Comment by razetime** [comment]
I tried running the following on my factor build in Ubuntu 20.04.5 LTS x86_64
```
<unrolled-list> 1 over push-back
               dup pop-back . dup pop-back .
```
It gives the expected `empty-deque` error after popping `1`.

**Comment by razetime** [comment]
@Kacarott was talking about a similar problem as well, but with pop-front:
```
<unrolled-list> 100 [1..b] [ over push-back ] each 100 [ dup pop-front drop ] times
```
`pop-front` and `pop-back` on the resulting deque both return 0.

**Comment by mrjbq7** [comment]
Awesome! Add a test case, push a fix?On Jan 31, 2023, at 7:32 AM, Raghu R ***@***.***> wrote:﻿
The problem is that when front and back in a list refer to the same array, and front-pos and back-pos are the same fixnum,   the list is empty. pop-front/existing and its back counterpart do not check for this and continues to keep pushing zeroes from the array. This is my rough patch:
: pop-front/existing ( list front -- )
    [ dup front-pos>> ] [ data>> ] bi* [ 0 ] 2dip set-nth-unsafe
    [ 1 + ] change-front-pos
    dup [ [ front>> ] [ back>> ] bi eq? ]
    [ [ front-pos>> ] [ back-pos>> ] bi = ] bi and
    [ dup clear-deque ] when
    drop ; inline

**Comment by razetime** [comment]
Fixed in [b44db33](https://github.com/factor/factor/commit/b44db33e3443595537167cd6284a1d060c002117).
